# Breast Cancer Prediction

In [22]:
#importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler , LabelEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.metrics import f1_score, classification_report, confusion_matrix, precision_score,accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

In [2]:
df=pd.read_csv('breast-cancer.csv')

### eda

In [3]:
df.head()

,id,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,...,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst
0,842302,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,842517,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,84300903,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,84348301,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,84358402,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


In [4]:
df.shape

(569, 32)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

In [5]:
df.columns

Index(['id', 'diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst'],
      dtype='object')

In [6]:
df.isnull().sum()

id                         0
diagnosis                  0
radius_mean                0
texture_mean               0
perimeter_mean             0
area_mean                  0
smoothness_mean            0
compactness_mean           0
concavity_mean             0
concave points_mean        0
symmetry_mean              0
fractal_dimension_mean     0
radius_se                  0
texture_se                 0
perimeter_se               0
area_se                    0
smoothness_se              0
compactness_se             0
concavity_se               0
concave points_se          0
symmetry_se                0
fractal_dimension_se       0
radius_worst               0
texture_worst              0
perimeter_worst            0
area_worst                 0
smoothness_worst           0
compactness_worst          0
concavity_worst            0
concave points_worst       0
symmetry_worst             0
fractal_dimension_worst    0
dtype: int64

In [11]:
df.head()

,id,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,...,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst
0,842302,1,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,842517,1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,84300903,1,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,84348301,1,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,84358402,1,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


# pre-processing data

In [10]:
#label encoding categorical feature
le=LabelEncoder()

df['diagnosis']=le.fit_transform(df['diagnosis'])

In [18]:
X=df.drop(columns=['id','diagnosis'])
y=df['diagnosis']

#train test split

X_train ,X_test,y_train,y_test=train_test_split(X, y , test_size=0.2 , random_state=42 , stratify=y)

In [14]:
preprocessor=ColumnTransformer(
    transformers=[("num",StandardScaler(),X.columns)]
    )


# Pipeline Setup

Each model have a seperate pipeline

In [48]:
# Machine learning models

lr=LogisticRegression()
dt=DecisionTreeClassifier(max_depth=7,min_samples_leaf=2)
rf=RandomForestClassifier(max_depth=7,min_samples_leaf=2)

In [49]:
#model pipelines

#logistic regression pipe
pipe_lr=Pipeline(
   steps=[
       ("preprocess",preprocessor),
       ("model",lr)
   ]
)

#decision tree pipe
pipe_dt=Pipeline(
   steps=[
       ("preprocess",preprocessor),
       ("model",dt)
   ]
)

#Random Forest pipe
pipe_rf=Pipeline(
   steps=[
       ("preprocess",preprocessor),
       ("model",rf)
   ]
)


In [56]:
from sklearn.ensemble import VotingClassifier

vrc=VotingClassifier(
    estimators=[("lr",pipe_lr),
                ("dt",pipe_dt),
                ("rf",pipe_rf)],
    voting="hard"
)

#Voting Classifier pipe
pipe_vrc=Pipeline(
   steps=[
       ("preprocess",preprocessor),
       ("model",rf)
   ]
)

## model training

In [57]:
models={"Logistic Regression":pipe_lr,
        "Decision Tree":pipe_dt,
        "Random Forest":pipe_rf,
        "Voting Classifier":vrc}

results=[]

for name,model in models.items():
    model.fit(X_train,y_train)
    y_train_pred=model.predict(X_train)
    y_test_pred=model.predict(X_test)

    #metrics
    train_acc=accuracy_score(y_train,y_train_pred)
    test_acc=accuracy_score(y_test,y_test_pred)
    precision = precision_score(y_test, y_test_pred, average='weighted')

    #stored in result
    results.append({
        "Model":name,
        "Train Accuracy":train_acc,
        "Test Accuracy":test_acc,
        "Precison Score":precision
    })

#converting dict -> dataframe
df_results=pd.DataFrame(results)

### results analysis

In [60]:
df_results

,Model,Train Accuracy,Test Accuracy,Precison Score
0,Logistic Regression,0.986813,0.964912,0.965185
1,Decision Tree,0.991209,0.938596,0.939001
2,Random Forest,0.993407,0.964912,0.966759
3,Voting Classifier,0.993407,0.973684,0.974737


# Key Insight

##### All models achieved strong performance, with test accuracies above 93%. Logistic Regression and Random Forest showed good generalization, while the Decision Tree displayed slight overfitting. The Voting Classifier outperformed individual models, achieving the highest test accuracy (97.4%) and precision (97.5%), confirming that ensemble methods provide more robust and reliable predictions. Overall, the results indicate that the ensemble approach is the most suitable candidate for deployment.